### DISCUSSION TITLE
bun.sh support

### DISCUSSION THREAD

**Discussion by saada** [remix-run/remix#5210] [Proposals]
### What version of Remix are you using?

❯ create-remix --version 1.6.5

### Steps to Reproduce

```sh
❯ bun --version
0.1.5
> bun run create-remix --template remix-run/indie-stack blog-tutorial
? TypeScript or JavaScript? TypeScript
? Do you want me to run `bun install`? Yes
⠦ Creating your app…error: Script not found "config"
Command failed: bun config get @remix-run:registry
error: Script not found "config"

error "create-remix" exited with 1 status
```

Looks like bun does not support the `config` command.

### Expected Behavior

I expected remix to just work.

### Actual Behavior

`bun config` does not exist as a command. We'll need to either skip that step or find another way?

Skipping this step and running `bun install` manually seems to be the workaround

```sh
? Do you want me to run `bun install`? No
```

**Comment by machour** [comment]
FYI, some work have been done (https://twitter.com/ebey_jacob/status/1544475779337048066) but no full support yet.

**Comment by djhi** [comment]
In case someone is wondering if it works now, I just started to use bun on a small remix project with tailwind, prisma & supabase among other things. Works flawlessly

*edit*: watch mode does not work

**Reply by MariaZentsova** [reply]
How did you install remix with bun? I run bun run create-remix and it's not working for me ..

**Comment by machour** [comment]
Converting this to a Proposal discussion, as per our [Development Process](https://remix.run/docs/en/v1/pages/contributing#development-process).

**Comment by electromidz** [comment]
Some one have suggestion for this issue ?! 
I want to create this remix with bun version 0.5.9 and I think have problem with that ...

**Comment by selbekk** [comment]
Now that Bun is in a stable 1.0 – what thoughts have been made to support running Remix on it?

**Reply by joakimbeng** [reply]
I guess it is possible using the [`--bun` flag](https://bun.sh/blog/bun-v0.8.0#support-for-nuxt-nuxt-dev)

**Reply by selbekk** [reply]
Not as a deployment target – you can generate Remix projects, but you can't use it as a drop-in replacement for Node.

**Reply by joakimbeng** [reply]
Why not you mean?

**Reply by redbmk** [reply]
@joakimbeng Not all the Node APIs are supported with `bun`. Some work would need to be done to make sure that remix is using Bun-native APIs where possible. For example:

```sh
❯ bunx remix dev

 💿  remix dev

 info  building...
 info  built (1.3s)
[remix-serve] http://localhost:3000 (...)

^C

❯ bunx --bun remix dev

 💿  remix dev

 info  building...
 info  built (1.3s)
50 |         }
51 |       }
52 |       return new FixedQueue;
53 |     }();
54 | 
55 |     function processTicksAndRejections() {
                                                                               ^
SyntaxError: Import named 'renderToPipeableStream' not found in module '.../node_modules/react-dom/server.browser.js'.
      at processTicksAndRejections (:55:76)
```

**Reply by redbmk** [reply]
Although that error is a bit confusing as it seems to be `react-dom` that's throwing the error, but Bun actually uses that same API in their example of React SSR

https://bun.sh/guides/ecosystem/ssr-react

**Reply by sergiodxa** [reply]
That specific error is not Remix, but React, the React team decided to only expose `renderToPipeableStream` on Node and `renderToReadableStream` on browser and Bun pick the browser bundle, the same happens with Cloudflare right now, if you check Cloudflare templates in `entry.server` is uses `renderToReadableStream` instead of `renderToPipeableStream`.

I made a repo where I used Bun to run my Remix app and it works this way too https://github.com/sergiodxa/remix-on-bun/blob/1f661192d759c7955b0884871024cada7a43cb20/app/entry.server.tsx#L13

A Remix app can run on Bun, the `dev` script that bundles has some issues but you can run the build with Node and the app with Bun, and on prod use only Bun.

You can also confirm that my repo uses Bun as the runtime because I'm using both Bun.serve and `bun:sqlite`.

**Reply by sergiodxa** [reply]
And on the Bun guide they also use `renderToReadableStream` instead of `renderToPipeableStream`

**Reply by redbmk** [reply]
Hmm, the error points to `server.browser.js` but there's actually a `server.bun.js`. There's also a `server.node.js`, so not sure why it's picking the browser one. Bun actually mentions React `18.3` includes some Bun native optimizations, but afaict the latest React is `18.2` and the PR that they reference isn't in that release.

It looks like if I install `react@next react-dom@next` it correctly fnids `server.bun.js` but I still get the same error

```sh
❯ NODE_ENV=production bunx --bun remix build
 info  building... (NODE_ENV=production)
 info  built (2.1s)
❯ bunx --bun remix-serve ./build/index.js
50 |         }
51 |       }
52 |       return new FixedQueue;
53 |     }();
54 | 
55 |     function processTicksAndRejections() {
                                                                               ^
SyntaxError: Import named 'renderToPipeableStream' not found in module '.../node_modules/react-dom/server.bun.js'.
      at processTicksAndRejections (:55:76)
```

Oh, and actually looking more closely I see the difference between the examples is `renderToPipeableStream` vs `renderToReadableStream`. For some reason the Pipeable version is only available in `node`, not `bun` or `browser`.

From what I can tell, Bun supports streams: https://bun.sh/docs/api/streams

> Bun also implements the node:stream module, including [Readable](https://nodejs.org/api/stream.html#stream_readable_streams), [Writable](https://nodejs.org/api/stream.html#stream_writable_streams), and [Duplex](https://nodejs.org/api/stream.html#stream_duplex_and_transform_streams). For complete documentation, refer to the [Node.js docs](https://nodejs.org/api/stream.html).

React's docs point to the same Node Stream page when talking about pipeable, so I'm not sure why React doesn't seem to support `renderToPipeableStream` for Bun. Maybe at least for that API that's more of a React issue

**Reply by redbmk** [reply]
OK, some more digging:


https://react.dev/reference/react-dom/server/renderToPipeableStream

> This API is specific to Node.js. Environments with [Web Streams,](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) like Deno and modern edge runtimes, should use [renderToReadableStream](https://react.dev/reference/react-dom/server/renderToReadableStream) instead.

I believe Remix supports edge runtimes, so they must be using ToReadableStraem in that case and would just need to do the same for Bun. Of course, that's just the first apparent issue, but at least fixing that would be a step towards full Bun support.

**Reply by redbmk** [reply]
I'm guessing the root of the fix will be creating a new template and runtime similar to deno: [template](https://github.com/remix-run/remix/tree/main/templates), [runtime](https://github.com/remix-run/remix/tree/main/packages/remix-deno)

**Reply by sergiodxa** [reply]
The problem is you're using Remix App Server, this one is using renderToPipeableStream because it's intended to work on Node, not Bun or Cloudflare.

So you need to change this in the entry.server.

**Reply by sergiodxa** [reply]
The main issue is that ReactDOM should expose `renderToReadableStream` to Node because ReadableStream is a thing now in Node, and that way the same code will work everywhere.

**Reply by eddyw** [reply]
If anyone (like me) finds themselves here:
```diff
-import { renderToPipeableStream } from "react-dom/server";
+import { renderToPipeableStream } from "react-dom/server.node";
```
In `entry.server.tsx` 😅

**Reply by danestves** [reply]
@eddyw from my side it keeps telling me that is using the browser version, what else did you do? 👀

**Reply by eddyw** [reply]
@danestves This is all things I changed:

*server.mjs*
```ts
import { createRequestHandler } from "@remix-run/express";
import { broadcastDevReady } from "@remix-run/node";
import express from "express";

let build = await import("./build/index.js");

const app = express();
app.use(express.static("public"));

const remix = createRequestHandler({
  build,
  mode: build.mode,
});

app.all("*", remix);

app.listen(3000, () => {
	if (process.env.NODE_ENV === "development") {
		broadcastDevReady(build);
	}
	console.log("App listening on http://localhost:3000");
});
```

*entry.server.tsx*
```diff
import { PassThrough } from "node:stream";
import type { AppLoadContext, EntryContext } from "@remix-run/node";
import { createReadableStreamFromReadable } from "@remix-run/node";
import { RemixServer } from "@remix-run/react";
// @ts-ignore
-import { renderToPipeableStream } from "react-dom/server";
+import { renderToPipeableStream } from "react-dom/server.node";
...
```

*package.json*
```diff
...
  "scripts": {
    "build": "NODE_ENV=production bunx --bun remix build",
    "build:dev": "NODE_ENV=development bunx --bun remix build",
    "start:bun": "NODE_ENV=production bun ./server.mjs",
    "start:node": "NODE_ENV=production node ./server.mjs",
    "dev": "bunx --bun remix dev"
  },
...
```

*remix.config.js*
```js
export default {
  ignoredRouteFiles: ["**/.*"],
  serverModuleFormat: "esm",
  dev: {
    manual: true,
    command: "bun --hot ./server.mjs",
  }
};
```

When in doubt, because prev build or cache has caused issues for me before when changing stuff (e.g entry server):
```bash
# rm -rf .cache/
# rm -rf build/
# bun dev
```

**Comment by maranomynet** [comment]
Really hope I can someday run Remix on Bun, without needing to have Node installed on the side, and without having to use a custom (entry.)server setup.  (I'm currently converting a project away from bun back to node+yarn+vitest to reduce the number of workarounds and engine jumps required, until things get resolved.)

@sergiodxa you talk about an issue with ReactDOM and that the dev build scripts are "mostly working".

Do you know if the React team have any plans to revolse their bit?

And is there any work planned on resolving the Remix dev build scripts?

**Reply by sergiodxa** [reply]
I don’t know if the React team is planning anything related to that.

For the dev script, now that the future will be Vite I’m not sure if you can run Vite with Bun.

**Reply by danestves** [reply]
Yes actually you can https://bun.sh/guides/ecosystem/vite

**Reply by maranomynet** [reply]
I noticed that `bun run --bun remix vite:dev` sort of runs, but has subtly different behavior from when I run it with node.

Client-side navigation waits for all `defer`red promises to resolve before rendering the new route, and some other minor things...

`bun run --bun remix vite:build` does absolutely nothing and exits with code 0.

(using bun v1.0.27 and remix v2.7.1)

**Reply by sergiodxa** [reply]
I don't think using Bun to build the app is a priority for the team, just using it to run the app. In theory since Bun has a bundler it should be possible to create a Bun plugin for Remix similar to the Vite plugin for Remix so you can directly use that, but then you will have to be sure to keep up with official changes to the Vite plugin to keep compatibility.

**Reply by maranomynet** [reply]
I see.
I was just surprised to find that the normal Vite plugin doesn't "just work" – since Bun aims to be drop-in compatible with Node.

So I don't think we should need a whole separate Bun vite plugin, rather to see what needs to tweaked so it doesn't break.
I can't tell if it's a Bun issue, or Remix/Vite issue.
Next.js supposedly "just works" on bun.

**Reply by sergiodxa** [reply]
> Next.js supposedly "just works" on bun

But what works? Does the compiler just work or the runtime? Because Remix runtime also just works, you only need Node for the build, but your app can run using Bun, I have tried it before, even the Bun HTTP server work with a simple adapter.

Also doing `bun run build` to run `next build` doesn't necessarily run with Bun, it's just using Bun as a replacement of npm, the only way to really be sure everything works is to don't have Node installed when running it.

**Reply by maranomynet** [reply]
If you add the `--bun` argument (after the `run` keyword) bun supposedly replaces the node shebang with the path to bun.

So in that case node should be out of the picture, unless the script spawns child processes that envoke node explicitly.

**Reply by maranomynet** [reply]
But you are right, the claim that Next works with bun, might indeed have this invisible asterisk next to it.

I haven't tried it myself. Been too busy playing with Remix. 😅

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
